multi/stm32l4-multi/rng.c

```diff
@@ -0,0 +1,160 @@
+/*
+ * Phoenix-RTOS
+ *
+ * STM32L4 RNG driver
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <board_config.h>
+#include <stdint.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/threads.h>
+#include <sys/platform.h>
+#include <sys/types.h>
+#include <phoenix/arch/stm32l4.h>
+
+#include "common.h"
+
+#define RNG_CR 0
+#define RNG_SR 1
+#define RNG_DR 2
+
+
+static struct {
+	volatile uint32_t *base;
+	handle_t lock;
+} rng_common;
+
+
+static int rng_readDR(uint32_t *val)
+{
+	int ret = -EAGAIN;
+	uint32_t sr = *(rng_common.base + RNG_SR);
+
+	if ((sr & 1u) != 0) {
+		*val = *(rng_common.base + RNG_DR);
+
+		/* Zero check is recommended in Reference Manual due to rare race condition */
+		ret = (*val != 0) ? 0 : -EIO;
+	}
+
+	/* Has error been detected? */
+	if ((sr & (3u << 5)) != 0) {
```

Comment:
Small findings:
1. In fact Clock Source Error don't require to reject data as faulty, according to RM, it's rather diagnostic information (it slows down the perypherium maybe?)
2. When something bad is happened with RNG, then it can be infinite loop in driver. Not sure if that's desirable

Reply:
1. I decided to reject the data in case of some deliberate attack on RNG entropy via influencing the clock. I disabled IP reinit in this case though,
2. added retry counter.